And /me and /authorization endpoints

jessebrennan opened the pull request:
This is an extension of #59.

Both #50 and #52 are covered in this PR because there was going to be some code overlap.

These changes are currently deployed at jesse.ucsc-cgp-dev.org

There were a couple outstanding issues that @coverbeck and I were discussing.

1. One is the access token in the /me endpoint. This is now removed.
2. Another was https://github.com/DataBiosphere/cgp-dashboard/compare/feature/commons...feature/51-auth-endpoint?expand=1#diff-3f41e546893dc64b71aaacad12cad815R357 
   I was wondering if comment is along the lines of what you were thinking.

jessebrennan committed:
Add /me endpoint that returns user's info

jessebrennan committed:
Update docstring

jessebrennan committed:
Improve error handling and messages

Now if token can't be refreshed this is explicitly logged and the
session is cleared.

If some other error happens a 401 is still returned, but message
reflects it better

jessebrennan committed:
Add auth endpoint (resolves #52)

jessebrennan committed:
Remove access token from /me endpoint

jessebrennan committed:
Reformat error message

coverbeck commented on app.py:
```diff
@@ -82,6 +82,8 @@ class ProdConfig(Config):
 login_manager.session_protection = "strong"
 
+# make a global bouncer instance to avoid needless re-instantiation
+whitelist_checker = Bouncer(os.getenv('EMAIL_WHITELIST_NAME'))
```
Will Bouncer handle getting `None` as a parameter?

jessebrennan replied:
Yes. The only problem with doing it this way is that future devs need to remember to check `EMAIL_WHITELIST_NAME` before making a call to Bouncer. Probably not a big deal.

coverbeck commented on app.py:
```diff
@@ -256,13 +258,19 @@ def parse_token():
         'client_id': Auth.CLIENT_ID,
         'client_secret': Auth.CLIENT_SECRET,
     }
     resp = oauth.refresh_token(Auth.TOKEN_URI, refresh_token=refresh_token, **extra)
```
I would like to see a comment here, that this line will raise the OAuth2Error. Either instead of or as well as your comment at the top. With just the comment at the top, it's not clear exactly where the error might get thrown, nor why you're not checking on line 273 that this line may have failed.

This is fairly nit-picky and personal taste, so feel free to ignore if you disagree.

jessebrennan replied:
I'll add another comment. I don't think they ever hurt.

coverbeck commented on app.py:
```diff
@@ -314,6 +322,132 @@ def check_session(cookie):
+        resp = _get_user_info_from_token()
+    # If there is a 5xx error, or some unexpected 4xx we will return the message but
+    # leave the token's intact b/c they're not necessarily to blame for the error.
+    assert resp.status_code == 200, str(resp.text)
```
I just read up on `assert` at https://stackoverflow.com/questions/5142418/what-is-the-use-of-assert-in-python, it being new to me in Python. According to that page, you can compile Python to ignore assert statements, so I'm wondering if you should handle this case without using assert?

Your call...

jessebrennan replied:
Yeah i personally don't like using asserts this way, I was just following precedence in other parts of the code. I'll change this one, and maybe a larger refactor / clean up can wait?

coverbeck replied:
Sounds good.

coverbeck commented on app.py:
```diff
@@ -314,6 +322,132 @@ def check_session(cookie):
+        }
+
+    In addition, if the access token was refreshed, the new access token
+    should be sent back in the session cookie.
```
`will be sent back`

coverbeck commented on app.py:
```diff
@@ -314,6 +322,132 @@ def check_session(cookie):
+    User is not authorized, return 403
+    """
+    if not os.getenv('EMAIL_WHITELIST_NAME'):
+        return '', 200
```
You should return 204 here, like you do on line 446.

jessebrennan replied:
See step 1 here https://github.com/DataBiosphere/cgp-dashboard/issues/52

If I should add a comment that makes me clear lmk.

coverbeck replied:
My bad on the original description in the issue... So I'm correcting myself. :)

If there is no content, which there isn't in this case, you're supposed to return 204, not 200.

coverbeck commented on app.py:
```diff
@@ -314,6 +322,132 @@ def check_session(cookie):
+    except OAuth2Error as e:
+        return 'Failed to get user info: ' + e.message, 401
+
+    output = dict((k, user_data[k]) for k in ('name', 'email'))
```
You need to check if user is still on whitelist

jessebrennan replied:
Whoops! duh...

coverbeck commented on app.py:
```diff
@@ -314,6 +322,132 @@ def check_session(cookie):
+    User is authorized, return 204
+    User is not authorized, return 403
+    """
+    if not os.getenv('EMAIL_WHITELIST_NAME'):
```
Instead of checking the env variable, then using the whitelist_checker, which means you have to guard against all uses of whitelist_checker by checking a different variable, maybe you could only set whitelist_checker if the variable is set at the top of the file, and then only check if whitelist_checker is set or not.

jessebrennan replied:
Ok, sure!

coverbeck commented on app.py:
```diff
@@ -314,6 +322,132 @@ def check_session(cookie):
+    If we get a working token, then ping google for user info, get
+    their email and check it against bouncer.
+
+    If there is no whitelist, return 200
```
204

jessebrennan replied:
https://github.com/DataBiosphere/cgp-dashboard/issues/52 see step 1

coverbeck commented on app.py:
```diff
@@ -508,11 +642,7 @@ def callback():
-            if whitelist:
-                b = Bouncer(whitelist)
-                if not b.is_authorized(email):
+            if os.getenv('EMAIL_WHITELIST_NAME') and not whitelist_checker.is_authorized(email):
```
See suggestions about just using one variable -- I would prefer something like the following (excuse any syntax errors), which seems clearer:

```
if whitelist_checker is not None and not whitelist_checker.is_authorized(email):
```

jessebrennan replied:
The closest thing to what you're describing would be

```
if whitelist_checker.secret_name is not None and not whitelist_checker.is_authorized(email):
```

But I think the current way is more clear than this. Do you disagree?

jessebrennan replied:
Nevermind, I see how this is connected to your previous comment

coverbeck replied:
I would not even initialize whitelist_checker itself if the environment variable is not set, and then do it like I suggested. 

My feeling is it is more straightforward to check if whitelist_checker is set and then use it if it is, vs. you checking an environment variable that you have to know the whitelist_checker has a dependency on and then using the whitelist_checker based on that. On top of that, you are creating an instance of the whitelist_checker when you don't really need it.

But this is getting really detailed for a very small thing and not worth it, so if you still disagree, I'm good with you doing it the current way.

coverbeck reviewed (changes requested):
Some of this is stylistic -- feel free to disagree with.

jessebrennan committed:
Implement review comments

- check for auth using whitelist_checker var
- add some comments
- change from AssertionError to ValueError
- return 204 instead of 200 if no authorization is required
- check if a suer is still whitelisted before sending back their
  info

coverbeck commented on app.py:
```diff
@@ -256,13 +261,20 @@ def parse_token():
         'client_id': Auth.CLIENT_ID,
         'client_secret': Auth.CLIENT_SECRET,
     }
+    # this call may throw an OAuth2Error
```
`Throw` sounds like Java, which I'm used to... :) Should be `raise` in Python? It's fine though.

jessebrennan replied:
I don't think anyone will misunderstand

coverbeck commented on app.py:
```diff
@@ -314,6 +326,133 @@ def check_session(cookie):
+    """
+    resp = _get_user_info_from_token(token=token)
+    if resp.status_code == 400:
+        # token expired, try once more
```
If token is not None, then they sent an Authorization header with the token... In that case, I think you should stop here -- not try to refresh the token, since you shouldn't have a refresh token anyway, nor do the session.pop stuff, because they've explicitly specified auth, so it shouldn't have the side effect of updating auth cookie values.

I suppose a user could have set the Authorization header AND have the cookie with refresh and access tokens, but it's unlikely, and if they do, I don't think we should mix the Auth header interpretation with the cookie values, if that make sense.

jessebrennan replied:
That makes sense. I only did it this way because I interpreted the docs to imply this. I'll change it now.

coverbeck commented on app.py:
```diff
@@ -314,6 +326,133 @@ def check_session(cookie):
+        return e.message, 401
+    except OAuth2Error as e:
+        return 'Failed to get user info: ' + e.message, 401
+    if not whitelist_checker.is_authorized(user_data['email']):
```
Do you need to check that whitelist_checker is not None?

jessebrennan replied:
Yes. Thanks!

coverbeck reviewed (changes requested):
One silly comment, one real thing, and one thing I think, but I'm not sure, that you have to address.

jessebrennan committed:
Add more review suggestions

When using a bearer token don't try to refresh the token in the
cookie.

also only only check the whitelist if necessary.

coverbeck reviewed (approved):
👍
